fair/src/autogluon/fair/utils/metrics.py

```diff
@@ -0,0 +1,125 @@
+import collections.abc
+import re
+import numpy as np
+
+from . import overall
+
+def fix_groups(per_group_metric,groups):
+    """fixes the choice of groups so that the metric can be passed to Pareto frontier algorithms"""
+    if isinstance(groups,collections.abc.Sequence):
+        masks = groups
+    else:
+        groups = np.asarray(groups)
+        unique = np.unique(groups)
+        masks = list()
+        for g in unique:
+            masks.append((groups==g))
+
+    def new_metric(y_true,y_est):
+        return per_group_metric(y_true,y_est,masks)
+    return new_metric
+
+
+def group_metric(metric,reduction=False):
+    """takes a metric as input and automatically generates a new metric that 
+    computes the metric per group and (optionally) applies a final reduction
+    e.g. group_metric(overall.recall) will compute the metric per group while
+    group_metric(overall.recall,'minimium') will return the lowest value of the metric on any group.
+    reduction may be one of 'min','max', 'max_diff', 'max_ratio' or a user specified function  """
+    if reduction is False:
+        reduction = lambda x: x
+    if reduction == 'min':
+        reduction = lambda x: x.min()
+    if reduction == 'max':
+        reduction = lambda x: x.max()
+    if reduction == 'max_diff':
+        reduction = lambda x: x.max()-x.min()
+    if reduction == 'max_ratio':
+        reduction = lambda x: x.max()/(1e-4+x.min())
+    
+    assert callable(reduction)
+
+    def new_metric(y_true, y_est, groups):
+        if isinstance(groups, collections.abc.Sequence):
+            masks = groups
+        else:
+            groups = np.asarray(groups)
+            unique = np.unique(groups)
+            masks = list()
+            for g in unique:
+                masks.append((groups==g))
+
+        out=np.zeros(len(masks))
+        for i, m in enumerate(masks):
+            out[i] = metric(y_true[m], y_est[m])
+        return reduction(out)
```

Comment:
If you are going to do this, please make it a class with a `__callable__` not a function. This will make life much better. Even better if that class happens to be an AutoGluon `Scorer`.